### DISCUSSION TITLE
Efficient rebuild of scripts when `lib` files change

### DISCUSSION THREAD

**Discussion by BeSpunky** [johnlindquist/kit#1247] [Share]
This script watches the `lib` folder, and when changes to `ts` files are made, it does 2 things:
1. Create/update a dependency graph of `libFilePath -> dependantScriptPaths[]`
2. Touches all script files that depend on the changed `lib` file to trigger rebuild.

https://gist.github.com/BeSpunky/d40bf6052b2c9a4f3a58b22108e5124a

**TLDR**
Currently, ScriptKit only rebuilds scripts if it detects changes to the `scripts` folder.
If you extract your reusable parts and put them into the `lib` folder, ScriptKit doesn't pick up on changes to those files.
The manual way to overcome this is to save your script file again and trigger rebuild.

No more... :)

Actually, this is a 3 scripts solution:
1. [`update-script-dependency-graphs`](https://gist.github.com/BeSpunky/c9139cdedfa349c501a70febea3c46d5): Partially rebuilds the graph if a triggering file has been provided, otherwise completely rebuilds it.
2. [`watch-libs`](https://gist.github.com/BeSpunky/d40bf6052b2c9a4f3a58b22108e5124a): Watches lib files, partially rebuild the graph using `update-script-dependency-graph`, then touch the scripts.
3. [`watch-scripts`](https://gist.github.com/BeSpunky/1ce1ad4e29b339bec11cd2d416cb676c): Watch script files, partially rebuild the graph

**Comment by nbbaier** [comment]
Could you give a little more guidance on how to install these? Would love to get it set up.

**Comment by BeSpunky** [comment]
Sure.
Do the following for the 3 Gists I've provided above:
1. Download the Gist as a zip.
Inside your zip, you'll notice two folders:

    ![image](https://user-images.githubusercontent.com/47897671/236613786-5dfde121-a87e-4dce-aa62-3099c5a2e27f.png)

2. Copy the contents of `_` into your `scripts` folder
3. Copy the contents of `__/lib` into your `lib` folder

> You might find duplicate lib files. Just skip the duplicates.
> That's because each script is generated separately.

Sorry for the hassle. I'm thinking of producing a quick installer script for these Gists.

@johnlindquist, any chance of leveraging the existing `kit://` schema for multiple files?
Otherwise, I'd have to come up with my own methodology for installation, AND... you'd have to install the installation script, before installing anything else. 🤓

**Comment by BeSpunky** [comment]
⚠️ HEADS UP

There's a problem I thought I had fixed, but turns out I didn't.
This might leave you with multiple unclosed processes over time, and you'll have to manually terminate them...

Will post the fix when I catch it.

**Comment by nbbaier** [comment]
Thanks! Having issues, might be something I did wrong though. I ge the following error when I add something to scripts:

```
[2023-05-06 13:04:38.082] [info]  {"args":["unlink"]}
[2023-05-06 13:05:33.866] [warn]  ☠️ ERROR PROMPT SHOULD SHOW ☠️
[2023-05-06 13:05:33.866] [warn]  Error: update-scripts-dependency-graph not found
    at resolveToScriptPath (file:///Users/nicholasbaier/.kit/core/utils.js:183:11)
    at run (file:///Users/nicholasbaier/.kit/core/utils.js:477:26)
    at file:///Users/nicholasbaier/.kenv/.scripts/watch-scripts.mjs:4:7
[2023-05-06 13:05:33.868] [info]  {"args":["add"]}
```

And the following error when I add to lib:

<img width="755" alt="Screenshot 2023-05-06 at 1 05 29 PM" src="https://user-images.githubusercontent.com/12950157/236640034-79632be0-0314-4678-b98c-3a605e35aaa5.png">

**Reply by BeSpunky** [reply]
A bit hard to tell...

What is your error message talking about users/nicholasbaier/lib/utils/package.json?
Is it part of you script?
Shouldn't that lib folder be under .kenv?

Also, try running the Update Script Dependency Graph script manually first.
You should have a new .kenv/db/_update-script-dependency-graph.json file. See what you get.

**Reply by nbbaier** [reply]
That's the weirdest thing. My lib/ folder is under .kenv! This is what the relevant part of my .kenv folder looks like after running the update script manually:

<img width="318" alt="Screenshot 2023-05-06 at 4 10 22 PM" src="https://user-images.githubusercontent.com/12950157/236646680-b9e92c57-adc1-45fe-bc7b-db786e56bb6a.png">

Here are the logs I get in watch-libs.log I get when moving a .ts file into and of my out of the libs folder
 
```
[2023-05-06 16:19:19.899] [info]  ../lib/utils
[2023-05-06 16:19:19.899] [info]  {"contents":"/Users/nicholasbaier/.kit/build/ts.js /Users/nicholasbaier/.kenv/scripts/watch-libs.ts"}
[2023-05-06 16:19:19.899] [warn]  ☠️ ERROR PROMPT SHOULD SHOW ☠️
[2023-05-06 16:19:19.899] [warn]  ReferenceError: runText is not defined
    at file:///Users/nicholasbaier/.kenv/.scripts/watch-libs.mjs:9:17
    at process.processTicksAndRejections (node:internal/process/task_queues:95:5)
[2023-05-06 16:19:19.903] [info]  {"args":["add"]}
[2023-05-06 16:19:39.591] [info]  ../lib/utils
[2023-05-06 16:19:39.591] [info]  {"contents":"/Users/nicholasbaier/.kit/build/ts.js /Users/nicholasbaier/.kenv/scripts/watch-libs.ts"}
[2023-05-06 16:19:39.592] [warn]  ☠️ ERROR PROMPT SHOULD SHOW ☠️
[2023-05-06 16:19:39.592] [warn]  ReferenceError: runText is not defined
    at file:///Users/nicholasbaier/.kenv/.scripts/watch-libs.mjs:9:17
    at process.processTicksAndRejections (node:internal/process/task_queues:95:5)
[2023-05-06 16:19:39.597] [info]  {"args":["unlink"]}
```
And here are the logs in update-scripts-dependency-graph.log when I move something into and out of my script folder.

```
[2023-05-06 16:21:48.822] [info]  👀 Dependency graph update triggered by '/Users/nicholasbaier/.kenv/scripts/test-file.ts'.
[2023-05-06 16:21:48.823] [info]  ⛓️ Performing a partial graph rebuild...
[2023-05-06 16:21:48.824] [info]  ✅ Done!
[2023-05-06 16:21:52.925] [info]  👀 Dependency graph update triggered by '/Users/nicholasbaier/.kenv/scripts/test-file.ts'.
[2023-05-06 16:21:52.925] [info]  ⛓️ Performing a partial graph rebuild...
[2023-05-06 16:21:52.926] [warn]  ☠️ ERROR PROMPT SHOULD SHOW ☠️
[2023-05-06 16:21:52.927] [warn]  Error: ENOENT: no such file or directory, open '/Users/nicholasbaier/.kenv/scripts/test-file.ts'
[2023-05-06 16:21:52.930] [info]  {"args":["unlink"]}
```

**Reply by BeSpunky** [reply]
What happens when script and lib files are already in place, and you just enter a file and save it?
Does it still break?

My current guess is that scripts my scripts can't handle broken/partial/unfinished operations too good.
Maybe that's also why the processes stay alive forever.

I'll make an attempt of wrapping some stuff with `try catch` tomorrow. See what comes out of it...

**Reply by nbbaier** [reply]
When I edited a script that was in scripts and saved I got no error, seems to have updated fine. I got the same error/logs as above for editing a ts file in libs.

**Reply by johnlindquist** [reply]
@BeSpunky @nbbaier 

The app will automatically handle this functionality in the next build (pushing it out late tonight).

I strongly advise against running these scripts in the latest build. They're definitely awesome learning resources though.

**Reply by BeSpunky** [reply]
Nice!
Would love to have it built in 💖

**Reply by JayAchar** [reply]
@johnlindquist Did this update get included? I've come across this issue today where editing files in a `lib` folder doesn't result in typescript rebuilding scripts.

**Reply by johnlindquist** [reply]
@JayAchar 

https://github.com/johnlindquist/kitapp/releases/tag/v1.93.13

☝️ There's the latest Script Kit v2 beta. I completely replaced the "compile on file change" logic into a custom loader that compiles instantly "on run". So no more lib/compile issues :)

**Reply by shyagamzo** [reply]
Wonderful! Thanks for the update!

**Reply by JosXa** [reply]
@johnlindquist In 2.9, it still only works if you touch the script file. It won't rebuild if its dependencies in lib change. I'll definitely grab this code and see what can be done with it :)

**Reply by johnlindquist** [reply]
@JosXa In 2.9, I'll need to clear the `.scripts` cache from the app if the `libs` change. The app no longer watches the scripts at all and uses a cache in the loader.
